### DISCUSSION TITLE
Event when object is changed with set or setOptions?

### DISCUSSION THREAD

**Discussion by radykal** [fabricjs/fabric.js#8866] [Q&A] [answered]
Is there any event triggered for objects when it has been changed via set or setOptions?

**Comment by iocoker** [comment, marked as answer]
@radykal Be careful hooking events to when an object is `set` as it can lead to an infinite loop if you also change`(set)` any property based on that event.
That being said, you can listen to the `object:modified` on the canvas, it covers most interactions. Others are `object:scaling, resizing, moving` etc.

To explicitly hook event to `set`, you may need to extend the object's class and trigger your event in `_set`.

**Comment by asturur** [comment]
Today there isn't such event and i m not sure is good for performances.
Fabric will try to use set more in the future in order to update object informations that otherwise needs to be recalculated every render.
Is there any specific use case why you would like to know when set is used?

If is an application need regarding user actions, while most of the internal set are handled already by existing events, for your own events you can just make your own set on the object prototype and call the standard set and fire an event.
But this would be just a central piece of your custom design that you can put higher up and doesn't need to be so much low lever in fabric inners

**Comment by ShaMan123** [comment]
This is resolved by using a proxy as we discussed
I still want that instead of set.
#8682

**Reply by asturur** [reply]
but that is too much, also my previous question is still valid. You generally do not want to trigger something every single property write, there must be something that you want to do that is related to what you want to build and doesn't have to go so much low level as property assignment.

**Reply by ShaMan123** [reply]
From experience of handling server updates, history management etc. there isn't a decent way of doing this because there are places that fabric doesn't use set, you must listen to numerous events, subclass, do additional state checks, blah blah, so it is not straight forward as I would want it to be.

**Reply by ShaMan123** [reply]
From experience of handling server updates, history management etc. there isn't a decent way of doing this because there are places that fabric doesn't use set, you must listen to numerous events, subclass, do additional state checks, blah blah, to cover all cases, so it is not straight forward as I would want it to be.
To answer your question, sometimes it isn't enough. Especially with complex UX. React integration also makes it less straight forward.
Personally I don't like to mix these things up. I prefer to subscribe to something (preferably one hook) and have all my server/history logic in one place for maintenance reasons and DX.
Using proxy hooks answers my needs (and is on the dev anyways in terms of perf).
Didn't you encounter these cases?
